### ISSUE TITLE
IDE protocol should provide underlying errors when disambiguation fails

### ISSUE THREAD

**Issue by mrmonday** [idris-lang/Idris-dev#2676]
The following code:

```idris
import Effects

foo : IO Int -> String -> IO Int
foo x = pure !x
```

Gives the error message:

```
Type checking ./foo.idr
foo.idr:4:5:When checking right hand side of foo:
Can't disambiguate since no name has a suitable type: 
        Effects.>>=, Prelude.Monad.>>=
Holes: Main.foo
```

Removing the import gives the expected error message:

```
Type checking ./foo.idr
foo.idr:5:1:When checking right hand side of foo:
When checking an application of function Prelude.Monad.>>=:
        Type mismatch between
                IO Int (Type of x)
        and
                (\uv => String -> uv) a (Expected type)
        
        Specifically:
                Type mismatch between
                        IO' (MkFFI C_Types String String)
                and
                        \uv => String -> uv
Holes: Main.foo
```

Which makes the solution, adding the missing parameter, far clearer.

**Comment by david-christiansen** [comment]
When you performed the import, you brought another `>>=` into scope. This means that both must be taken into account. The error message that you expected here would be wildly misleading if you'd expected to use the `>>=` from `Effects`, after all.

I suppose one reasonable improvement to make would be to make the error clickable or expandable and then optionally show the error that prevented each overloading of `>>=` from working.

**Comment by mrmonday** [comment]
I understand why that error occurs, it's not particularly useful though. Seeing the errors for why each version doesn't work would likely solve the issue of the "poor error message", I think. That solution would likely solve another error message issue which I haven't managed to reduce into a concise test case too.

**Comment by david-christiansen** [comment]
In the meantime, you can explicitly qualify it to the one you mean and then get the specific error.

**Comment by mrmonday** [comment]
That would be what I usually do for this kind of error - my actual issue was nestled in a do block, meaning I'd have to rewrite the entire thing to achieve it, unfortunately. Thank you.

**Comment by edwinb** [comment]
The intention here is that you explicitly qualify (or use a 'with') and see what happens. If we were to show the error messages for all versions, we'd have differently bad error messages which gave too much information, especially when there are a lot of possibilities.

While it's clearly desirable for error messages to tell you what you need to know to fix the problem, Idris can't always guess what you're trying to do (it used to try, and it was quite bad at it). So instead, it tries to tell you exactly what's wrong.

I'm going to close this, because I've considered other ways of presenting this error message in the past (including the suggested way) and I still think that what the REPL does is the right default. It would be good for an interactive tool (e.g the emacs mode) to have a clickable way to expand the error, both for this and for other kinds of errors, but I think that is a different matter.

**Comment by david-christiansen** [comment]
@edwinb I agree that the current error message is good for the REPL, but that the IDE protocol should provide the original error messages so that the user can expand them in place. Do you mind if I leave this open for that? I think it has useful context and examples.

**Comment by edwinb** [comment]
I think it's a different issue, really. I don't mind it being kept open if the title and description reflects the actual issue though...

**Comment by ahmadsalim** [comment]
@david-christiansen I am not sure I understand the issue, should it still be open?
